### ISSUE TITLE
📎 handle advanced generic use cases

### ISSUE THREAD

**Issue by arendjr** [biomejs/biome#6609]
### Description

This is quite a crazy use case, but with better generics handling, we should be able to squeeze two diagnostics out of this one:

```ts
function createLazyPromise<
	T extends string,
	U extends (prop: PropertyKey) => Promise<T>,
>(getValue: U) {
	let resolve: (value: T) => void;
	const promise = new Promise<T>((r) => {
		resolve = r;
		return r;
	});

	const proxy = new Proxy(promise as Promise<T>, {
		get(target, prop, receiver) {
			if (prop in target) {
				return Reflect.get(target, prop, receiver);
			}
			// First diagnostic expected:
			getValue(prop).then(resolve);
			return undefined; // Could also return another proxy here
		},
	});

	return proxy as Promise<T>;
}

const lazy = createLazyPromise((prop) =>
	Promise.resolve(`You accessed: ${String(prop)}`),
);

// Second diagnostic expected:
lazy.then((result) => {
	console.log(result); // "You accessed: foo"
});
```
